### DISCUSSION TITLE
Size of PDF file generated is high

### DISCUSSION THREAD

**Discussion by shashikulk** [pymupdf/PyMuPDF#2987] [Q&A] [answered]
First of all, thanks for adding ```insert_htmlbox()``` feature/function to pymupdf. It works very well for non-English script. However, I have one observation regarding file size.

I am using ```insert_htmlbox()``` to insert some text (English as well as Devanagari) to my file. The size of the file created seems very high for the amount of text content I am writing to it. Below is the code I am using as an example in which I am writing only 1 line of Devanagari text to it, and the size of the pdf file generated is 242 KB. 

```
import fitz_new as fitz
import os
out_path = str(os.getcwd()) + '/output/'
pdffilename = out_path + 'sample.pdf'

doc = fitz.open()
page = doc.new_page()
text = "PDF तयार करण्याचा हा माझा पहिला दस्तऐवज प्रकल्प आहे"
rect = fitz.Rect(30, 30, 480, 55)
page.insert_htmlbox(rect, text)
doc.save(pdffilename)
doc.close()
```
If I add more lines of text using insert_htmlbox(), the file size increases proportionately. Is there any way to keep the pdf file size to low?

I am using latest pymupdf with Python 3.10.x and Win11.

regards,
Shashikant

**Comment by JorjMcKie** [comment, marked as answer]
The method dynamically pulls in all required fonts. These may be several. Asian fonts typically are large, but the number of glyphs that you actually use in your text is small.
Therefore you should make sure to
* build font subsets which only contain the glyphs that are actually in use
* save the file with all available compression options

Please try the following code version:
```python
import fitz_new as fitz
import os
out_path = str(os.getcwd()) + '/output/'
pdffilename = out_path + 'sample.pdf'

doc = fitz.open()
page = doc.new_page()
text = "PDF तयार करण्याचा हा माझा पहिला दस्तऐवज प्रकल्प आहे"
rect = fitz.Rect(30, 30, 480, 55)
page.insert_htmlbox(rect, text)
doc.subset_fonts()
doc.ez_save(pdffilename)
doc.close()
```

**Comment by JorjMcKie** [comment]
Going to move this to the "Discussions" tab.

**Reply by shashikulk** [reply]
Thanks for the solution. It worked perfectly to reduce the file size.

**Reply by JorjMcKie** [reply]
Hope you will enjoy this new function. In the next version (rolled out very soon), we are also introducing an additional parameter "opacity". This will make the content in the rectangle transparent ... may also be an interesting option ...

**Reply by shashikulk** [reply]
Ok. Will try it out once it is rolled out. 
Meanwhile, I would like to know if it is possible for the text within a single htmlbox to be selectively made bold/italic. For example, can we make the word 'दस्तऐवज' bold in the sentence "PDF तयार करण्याचा हा माझा पहिला दस्तऐवज प्रकल्प आहे", without having to use separate htmlbox for that word and rest of the sentence?

**Reply by JorjMcKie** [reply]
> Meanwhile, I would like to know if it is possible for the text within a single htmlbox to be selectively made bold/italic. For example, can we make the word 'दस्तऐवज' bold in the sentence "PDF तयार करण्याचा हा माझा पहिला दस्तऐवज प्रकल्प आहे", without having to use separate htmlbox for that word and rest of the sentence?

Of course this is possible!
Please have a look [here](https://pymupdf.readthedocs.io/en/latest/recipes-text.html#how-to-fill-a-box-with-html-text).
You are completely free to do whatever your HTML knowledge allows you to do 😉.
However, whether or not your Devanagari text will react to bold / italic will depend on the font(s) you are supplying. Colors are no problem.
MuPDF by default pulls in Noto fonts only - which not always contain versions for bold / italic.
But if you have your own full set of regular, bold, italic and bold-italic version of a Devanagari font, then this is perfectly possible too, see an example [here](https://pymupdf.readthedocs.io/en/latest/recipes-text.html#how-to-specify-your-own-fonts).

**Reply by JorjMcKie** [reply]
And as documented:
You can choose between either always fitting the content in the given rectangle (using shrink when necessary), or accepting a no-op if content does not fit.

**Reply by shashikulk** [reply]
Thanks a lot. Will try these methods / features. May have to learn a little bit of HTML ;-)

**Reply by shashikulk** [reply]
Need some help.
I am trying to create a table in PDF using the approach given [here](https://pymupdf.readthedocs.io/en/latest/recipes-text.html#how-to-output-html-tables-and-images)
It is working fine, except a few styles (table border radius and column widths) are not reflecting in the output pdf. If I view the html (the contents of ```text``` variable I am generating through my code) in a browser, all the styles are showing correctly. What could be the reason for difference in behavior in pdf output created by ```insert_htmlbox``` vs. browser? Or, am I missing something?

Given below is my code for reference.

```
import random
import fitz_new as fitz

srno, ques, ans = [], [], []
filename = "<your-local-file-name>.pdf"
qlang = random.choice(['English', 'Marathi'])
print('language is ', qlang)
for i in range(15):
    srno.append(str(i+1))
    if qlang == 'English':
        ques.append("random text # This is some random text "+str(i+1))
    else:
        ques.append("न्यूयॉर्क, श्रावणबेळगोळा, कल्याण # यादृच्छिक प्रश्न " + str(i + 1))
    ans.append(str(random.randint(1000, 10000000)))

if qlang == 'English':
    mystyle = """<td style="font-family: QuicksandFM;">"""
else:
    mystyle = """<td style="font-family: NotoSansDevanagari;">"""

arch_OT = fitz.Archive('<your-local-fonts-directory>')
doc = fitz.Document()
page = doc.new_page()

text = """
<style>
@font-face {font-family: NotoSansDevanagari; src: url(NotoSansDevanagari-Regular.ttf);}
@font-face {font-family: QuicksandFM; src: url(Quicksand-Regular.ttf);}
@font-face {font-family: QuicksandFM; src: url(Quicksand-Bold.ttf); font-weight: bold;}
td, th {padding: 5px; text-align: center;}
table {
    border: 0.25px solid blue;
    border-spacing: 0;
    border-radius: 2px;
    table-layout: fixed;
}
table td, table th {
    border: 0.25px solid black;
}
table tr td {
    border-right: 0;
    border-bottom:0;
}
table th {
    border-right: 0;
    border-top: 0;
    border-bottom: 0;
    background-color: #E0FFFF;
}
</style>
<body>
<table>
    <tr style="font-family: QuicksandFM;">
        <th style="border-left:0; width: 30px;"><b>#</b></th>
        <th style="width: 600px;"><b>Question</b></th>
        <th style="width: 100px;"><b>Answer</b></th>
    </tr>
"""
for i in range(15):
    text = text + """<tr>
        <td style="border-left:0; font-family: QuicksandFM;">""" + srno[i] + """</td>
        """ + mystyle + ques[i] + """</td>
        <td style="font-family: QuicksandFM;">""" + ans[i] + """</td>
        </tr>"""

text = text + """
    </table>
    </body>
    """

print('text is ', text)
rect = fitz.Rect(30, 30, 680, 700)
page.insert_htmlbox(rect, text, archive=arch_OT)
doc.ez_save(filename)

```

**Reply by JorjMcKie** [reply]
Column widths of HTML tables cannot be influenced - at least not easily. They are computed and optimized by the Story code itself.
You **_can_** exert some tricky influence by adding a table row with font height 0 and a desired number of non-breaking spaces `&nbsp;` - normal double spaces are contracted to one space as usual.
Not sure about border radius yet. Maybe that it indeed is unsupported by MuPDF. I can ask if you need it or ask yourself in the public MuPDF [Discord](https://discord.gg/mbdSPq9a) channel. MuPDF Story supports HTML4 and CSS2 roughly.

Other comment:
You shouldn't need to add font NotoSansDevanagari-Regular.ttf - it is one of the MuPDF standard ones. But won't harm either to name it that way.

**Reply by JorjMcKie** [reply]
Ah - and you definitely should use `doc.subset_fonts()` immediately before save.
Also make sure to use garbage collection 3 or simply ez_save.

**Reply by JorjMcKie** [reply]
All Asian fonts are large and will thus benefit very much from building font subsets (= remove unused glyphs).

**Comment by shashikulk** [comment]
Thanks for your response. 

Since column width and row height are not in my control as a user of ```insert_htmlbox()```, and the ```&nbsp;``` approach was getting trickier due to varying length of text in each row, I decided to use ```make_table()``` individually for each column.

Now I am using ```make_table()```, coupled with ```shape.draw_rect()``` and ```insert_htmlbox()```. So far it is giving me desired result and control in terms of column width and row height.  

Noted your other comments.
